### PR TITLE
client: reconnect after blacklisted

### PR DESCRIPTION
**Pull request by ukernel**
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

File: src/client/Client.cc [hunk 1]
```diff
@@ -4142,6 +4153,17 @@ void Client::remove_session_caps(MetaSession *s)
...
     }
+    caps &= CEPH_CAP_FILE_CACHE | CEPH_CAP_FILE_BUFFER;
+    if (err == -EBLACKLISTED &&
+	cct->_conf.get_val<bool>("client_discard_dirty_on_blacklisted") &&
+	caps && !in->caps_issued_mask(caps, true)) {
+      if (in->oset.dirty_or_tx) {
+	lderr(cct) << __func__ << " still has dirty data on " << *in << dendl;
+	in->set_async_err(err);
+      }
+      objectcacher->purge_set(&in->oset);
```
Comment (batrick) [inline src/client/Client.cc]: What happens when a future `fsync(fd)` comes? I may be missing it, but we would want `-EIO` in that case. Yes?

Comment (ukernel) [reply src/client/Client.cc]: the set_async_err() above make future fsync/close return error.

**Comment by tchaikov** [comment]
retest this please

**Comment by batrick** [comment]
retest this please

**Comment by batrick** [comment]
passed QA, just waiting for test

**Comment by batrick** [comment]
ping @ukernel

**Comment by batrick** [comment]
retest this please

File: qa/tasks/cephfs/test_client_recovery.py [hunk 2]
```diff
@@ -544,3 +544,28 @@ def test_abort_conn(self):
         self.assert_session_state(gid, "open")
         time.sleep(session_timeout * 1.5)  # Long enough for MDS to consider session stale
         self.assert_session_state(gid, "stale")
+
+    def test_reconnect_after_blacklisted(self):
+        """
+        Check that client uses new entity addr to reconnect after blacklisted
+        """
+        if not isinstance(self.mount_a, FuseMount):
+            raise SkipTest("Require FUSE client to handle signal STOP/CONT")
```
Comment (batrick) [inline qa/tasks/cephfs/test_client_recovery.py]: Should say kernel doesn't yet support this.

File: qa/tasks/cephfs/test_client_recovery.py [hunk 3]
```diff
@@ -544,3 +544,28 @@ def test_abort_conn(self):
...
+
+        # Ask client to reconnect
+        self.mount_a.admin_socket(['kick_stale_sessions'])
+
+        # Make sure operation does not hang
+        self.mount_a.run_shell(["touch", "testfile"])
+
+        self.mount_a.gather_mount_info();
+        self.assertFalse(self.mount_a.is_blacklisted())
+        self.assert_session_count(2)
```
Comment (batrick) [inline qa/tasks/cephfs/test_client_recovery.py]: It'd also be good to test that an in-progress write fails with EIO. Using self.write_background should work.

**Comment by batrick** [comment]
Note to self: run with current version: /ceph/teuthology-archive/pdonnell-2019-05-15_13:28:11-fs-wip-pdonnell-testing-20190515.015737-distro-basic-smithi/3956764/teuthology.log

**Comment by ukernel** [comment]
I don't like to let client auto reconnect after blacklisted. Otherwise client with 'client_reconnect_stale == true' can't be blacklisted.  User need to explicitly run 'kick_stale_sessions' asok command.

**Comment by batrick** [comment]
> I don't like to let client auto reconnect after blacklisted. Otherwise client with 'client_reconnect_stale == true' can't be blacklisted. User need to explicitly run 'kick_stale_sessions' asok command.

So you're concerned about a client that cannot be permanently evicted? Don't we have the ability to blacklist an entire IP address if the admin wants to permanently get rid of it?

**Pull request closed without merging**